### PR TITLE
RTDETR update

### PR DESCRIPTION
**Pull request by shijianjian**
#### Changes

Fixes # (issue)

#### Type of change

- [ ] 📚  Documentation Update
- [ ] 🧪 Tests Cases
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🔬 New feature (non-breaking change which adds functionality)
- [ ] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📝 This change requires a documentation update

#### Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

File: kornia/contrib/object_detection.py [hunk 1]
```diff
@@ -113,18 +112,18 @@ def __init__(self, size: tuple[int, int], interpolation_mode: str = "bilinear")
...
 
-    def forward(self, imgs: list[Tensor]) -> tuple[Tensor, list[ImageSize]]:
+    def forward(self, imgs: list[Tensor]) -> tuple[Tensor, Tensor]:
         # TODO: support other input formats e.g. file path, numpy
         resized_imgs, original_sizes = [], []
         for i in range(len(imgs)):
             img = imgs[i]
             # NOTE: assume that image layout is CHW
-            original_sizes.append(ImageSize(height=img.shape[1], width=img.shape[2]))
+            original_sizes.append([img.shape[1], img.shape[2]])
```
Comment (edgarriba) [inline kornia/contrib/object_detection.py:121]: can you specify in the docs the return order of this. This is usually the root of bugs.

File: kornia/contrib/models/rt_detr/architecture/rtdetr_head.py [hunk 2]
```diff
@@ -192,14 +193,12 @@ def forward(
         return out
 
 
-class TransformerDecoder:
-    def __init__(self, hidden_dim: int, decoder_layers: nn.ModuleList, num_layers: int, eval_idx: int = -1) -> None:
+class TransformerDecoder(Module):
+    def __init__(self, hidden_dim: int, decoder_layer: nn.Module, num_layers: int, eval_idx: int = -1) -> None:
         super().__init__()
-        self.layers = decoder_layers
+        # self.layers = decoder_layers
```
Comment (edgarriba) [inline kornia/contrib/models/rt_detr/architecture/rtdetr_head.py:199]: remove commented code

File: kornia/contrib/models/rt_detr/model.py [hunk 3]
```diff
@@ -167,10 +233,10 @@ def forward(self, images: Tensor) -> tuple[Tensor, Tensor]:
               :math:`K` is the number of classes.
             - **boxes** - Tensor of shape :math:`(N, Q, 4)`, where :math:`Q` is the number of queries.
         """
-        if self.training:
-            raise RuntimeError("Only evaluation mode is supported. Please call model.eval().")
+        # if self.training:
```
Comment (edgarriba) [inline kornia/contrib/models/rt_detr/model.py:236]: remove ? or put comment if we want to support that later

File: kornia/contrib/models/rt_detr/post_processor.py [hunk 4]
```diff
@@ -38,8 +38,8 @@ def forward(self, logits: Tensor, boxes: Tensor, original_sizes: list[ImageSize]
         boxes_xy = concatenate([cxcy - wh * 0.5, wh], -1)
 
         sizes_wh = torch.empty(1, 1, 2, device=boxes.device, dtype=boxes.dtype)
-        sizes_wh[..., 0] = original_sizes[0].width
-        sizes_wh[..., 1] = original_sizes[0].height
+        sizes_wh[..., 0] = original_sizes[0][0]
+        sizes_wh[..., 1] = original_sizes[0][0]
```
Comment (edgarriba) [inline kornia/contrib/models/rt_detr/post_processor.py:42]:
```suggestion
        sizes_wh[..., 1] = original_sizes[0][1]
```
you made your own bug here :)

**Review by edgarriba — APPROVED** [review]

File: kornia/contrib/models/rt_detr/model.py (hunk 3 above)
Comment (shijianjian) [reply kornia/contrib/models/rt_detr/model.py:236]: IDK why we put it. It can be optimised if I write the boiler code by hand.

Comment (edgarriba) [reply kornia/contrib/models/rt_detr/model.py]: Then remove it

File: docs/source/models/rt_detr.rst [hunk 5]
```diff
@@ -1,6 +1,28 @@
...
+    detector = RTDETRDetectorBuilder.build("rtdetr_r18vd", image_size=640)
+
+    # get the output boxes
+    boxes = detector(input_img)
+
+    # draw the bounding boxes on the images directly.
+    output = detector.draw(input_img, output_type="pil")
+    output[0].save("Kornia-RTDETR-output.png")
+
+    # convert the whole model to ONNX directly
```
Comment (edgarriba) [inline docs/source/models/rt_detr.rst:22]: I feel that this object should be able to be passed to the Onnx sequential stuff

File: kornia/models/detector/rtdetr.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
+import warnings
+from typing import Optional
+
+import torch
+from torch import nn
+
+from kornia.contrib.models.rt_detr import DETRPostProcessor
```
Comment (edgarriba) [inline kornia/models/detector/rtdetr.py:7]: Why don’t we move it in this new module ?

File: kornia/models/detector/rtdetr.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    This class provides static methods to:
+        - Build an object detection model from a model name or configuration.
+        - Export the model to ONNX format for inference.
+    """
+
+    @staticmethod
+    def build(
+        model_name: Optional[str] = None,
+        config: Optional[RTDETRConfig] = None,
+        pretrained: bool = True,
```
Comment (edgarriba) [inline kornia/models/detector/rtdetr.py:25]: Tuple ?

**Review by edgarriba — COMMENTED** [review]
Have you checked how much this differs from v2? https://github.com/lyuwenyu/RT-DETR/tree/main/rtdetrv2_pytorch

File: docs/source/models/rt_detr.rst (hunk 5 above)
Comment (shijianjian) [reply docs/source/models/rt_detr.rst:22]: It definitely can.

File: kornia/models/detector/rtdetr.py (hunk 6 above)
Comment (shijianjian) [reply kornia/models/detector/rtdetr.py:7]: Lots of things will break, some testing or maybe documentation. I do not want to touch it for now.

File: kornia/models/detector/rtdetr.py (hunk 7 above)
Comment (shijianjian) [reply kornia/models/detector/rtdetr.py:25]: what?

Comment (edgarriba) [reply kornia/models/detector/rtdetr.py:25]: sorry my bad, that was supposed to be in another line where image size is only defined by a single int

File: kornia/contrib/models/rt_detr/post_processor.py [hunk 8]
```diff
@@ -61,10 +61,8 @@ def forward(self, logits: Tensor, boxes: Tensor, original_sizes: Tensor) -> Tens
         cxcy, wh = boxes[..., :2], boxes[..., 2:]
         boxes_xy = concatenate([cxcy - wh * 0.5, wh], -1)
 
-        sizes_wh = torch.empty(1, 1, 2, device=boxes.device, dtype=boxes.dtype)
-        sizes_wh[..., 0] = original_sizes[0][1]
-        sizes_wh[..., 1] = original_sizes[0][0]
-        sizes_wh = sizes_wh.repeat(1, 1, 2)
+        # Get dynamic size from the input tensor itself
```
Comment (edgarriba) [inline kornia/contrib/models/rt_detr/post_processor.py:64]: let's add a todo or open ticket to update with a conversion free function

File: kornia/contrib/models/rt_detr/post_processor.py [hunk 9]
```diff
@@ -78,4 +76,7 @@ def forward(self, logits: Tensor, boxes: Tensor, original_sizes: Tensor) -> Tens
         index = index // num_classes
         boxes = boxes_xy.gather(dim=1, index=index.unsqueeze(-1).repeat(1, 1, boxes_xy.shape[-1]))
 
-        return concatenate([labels[..., None], scores[..., None], boxes], -1)
+        all_boxes = concatenate([labels[..., None], scores[..., None], boxes], -1)
+
+        return all_boxes[(all_boxes[:, :, 1] > self.confidence_threshold).unsqueeze(-1).expand_as(all_boxes)].view(
```
Comment (edgarriba) [inline kornia/contrib/models/rt_detr/post_processor.py:81]: exclude this for onnx export

File: kornia/contrib/object_detection.py [hunk 10]
```diff
@@ -119,22 +119,26 @@ def __init__(self, size: tuple[int, int], interpolation_mode: str = "bilinear")
...
+
+        iters = len(imgs) if isinstance(imgs, list) else imgs.shape[0]
+        original_sizes = imgs.new_zeros((imgs.shape[0], 2))
+        for i in range(iters):
             img = imgs[i]
-            original_sizes.append([img.shape[-2], img.shape[-1]])
+            original_sizes[i, 0] = img.shape[-2]  # Height
+            original_sizes[i, 1] = img.shape[-1]  # Width
             resized_imgs.append(
                 # TODO: fix kornia resize warnings
```
Comment (edgarriba) [inline kornia/contrib/object_detection.py:138]: maybe remove this ? if warning are fixed

File: kornia/contrib/object_detection.py [hunk 11]
```diff
@@ -202,13 +212,14 @@ def save(self, images: list[Tensor], directory: Optional[str] = None) -> None:
...
             directory = os.path.join("Kornia_outputs", name)
-        outputs = self.draw(images)
+        outputs = self.draw(images, detections)
         os.makedirs(directory, exist_ok=True)
         for i, out_image in enumerate(outputs):
             write_image(
                 os.path.join(directory, f"{str(i).zfill(6)}.jpg"),
                 out_image.mul(255.0).byte(),
             )
+        print(f"Outputs are saved in {directory}")
```
Comment (edgarriba) [inline kornia/contrib/object_detection.py:222]: we should ne using the `logging` library for this

File: kornia/core/external.py [hunk 12]
```diff
@@ -24,6 +27,10 @@ def __init__(self, module_name: str) -> None:
         self.module_name = module_name
         self.module: Optional[ModuleType] = None
 
+    def _install_package(self, module_name: str) -> None:
+        print(f"Installing `{self.module_name}` ...")
```
Comment (edgarriba) [inline kornia/core/external.py:31]: use `logging`

File: kornia/contrib/models/rt_detr/post_processor.py (hunk 8 above)
Comment (shijianjian) [reply kornia/contrib/models/rt_detr/post_processor.py:64]: it is done here. Just a note

Comment (edgarriba) [reply kornia/contrib/models/rt_detr/post_processor.py:64]: I mean something to be used from Kornia.geometry.boxes

**Review by edgarriba — APPROVED** [review]
LGTM. Test on Windows are failing because of no numpy.

**Pull request merged**